### ISSUE TITLE
Verify spire-agent high cpu load caused by k8s JSON parsing

### ISSUE THREAD

**Issue by szvincze** [spiffe/spire#3957]
* **Version**: 1.5.2
* **Platform**: Linux 5.19.0-35-generic_#36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC x86_64
* **Subsystem**: agent

During performance tests we experienced that spire-agents utilize high CPU load while spire-servers hardly utilize CPU.

```
NAME                CPU(cores)   MEMORY(bytes)
spire-agent-5nlk4   488m         47Mi
spire-agent-c7tpr   634m         51Mi
spire-agent-jgsvr   464m         58Mi
spire-agent-l9zzz   406m         39Mi
spire-agent-mdstq   487m         60Mi
spire-agent-qnpd2   390m         45Mi
spire-agent-r6r94   509m         40Mi
spire-server-0      7m           117Mi
spire-server-1      3m           48Mi
```
There are 46 registration entries with 1h TTL. There are 4 worker nodes in the system, the maximum limit is 500 pods/worker node. During the test there are ~70 pods running per worker nodes.

Below I copied the agent configuration, which is very simple:
```
agent {
    data_dir = "/run/spire/temp"
    log_level = "DEBUG"
    server_address = "spire-server"
    server_port = "8081"
    socket_path = "/run/spire/sockets/agent.sock"
    trust_bundle_path = "/run/spire/bundle/bundle.crt"
    trust_domain = "example.org"
}

plugins {
    NodeAttestor "k8s_psat" {
        plugin_data {
            cluster = "example-cluster"
        }
    }

    KeyManager "memory" {
        plugin_data {}
    }

    WorkloadAttestor "k8s" {
        plugin_data {
            skip_kubelet_verification = true
        }
    }
}
```
By default the TTL was 1h for SVIDs and 12h for the CA but we tried to increase and decrease them without significant effect on the CPU utilization.

Based on our slack discussion, we tried to add batches of 20 pods on the cluster in 11 steps. The below diagram shows how the CPU utilization of the spire-agent pod increased in time:
![cpu_usage_agent_new_pods_iteratively](https://user-images.githubusercontent.com/90055916/224092050-eae6e624-760a-43a3-b2d0-59bc833a70aa.png)
The utilization is quite linear with the increasing number of running pods. We experienced the same behavior on each worker nodes, so in the peak period spire-agent utilized ~15% of the CPU. Then it was broken down when deleted all the pods we created previously.
In the meantime the spire-agent logs don't contain any suspicious activity that would show the agents would be busy at all.

We managed to reproduce it on more clusters however the CPU utilization was different which we cannot explain.
In understand that it depends on the amount of the running pods, but do all the processes running on that worker node count?

Do you have experience with crowded clusters with many pods, how spire-agent's CPU utilization looks like?
Is this the expected behavior?

**Comment by azdagron** [comment]
It might help if you enable the profiling endpoints on the SPIRE agent and collect some pprof profiles for the cpu. We can then figure out where the time is being spent.

See the `profiling*` configurables under https://github.com/spiffe/spire/blob/main/doc/spire_agent.md#agent-configuration-file.

**Comment by szvincze** [comment]
@azdagron, I am trying to create a profile for the CPU in the same way as a year ago, but if I use the `profiling_names = ["cpu"]` I get this error message when capturing the profile:
```
http://localhost:1234/debug/pprof/profile?seconds=30: server response: 500 Internal Server Error - Could not enable CPU profiling: cpu profiling already in use
```
So, I just set the `profiling_enabled = true` and `profiling_port` in the agent.conf, specify the 60 seconds as duration and capture the profile like this: `curl -s http://localhost:1234/debug/pprof/profile?seconds=60 > cpu.out`

Is the 60 seconds duration enough?

**Comment by azdagron** [comment]
Hmm, strange. Anyway, yes, 60s should be plenty if the process is under the load you expect.

**Comment by szvincze** [comment]
@azdagron: I attached the [output](https://github.com/spiffe/spire/files/10943364/cpu.out.gz) of the profiling.

**Comment by azdagron** [comment]
The profile shows the time spent parsing the pod list returned from the kubelet during K8s workload attestation. I think this would imply that workloads are attesting, a lot.

Do you see anything in the logs for these workloads? Are they spinning connecting to the Workload API?

> In the meantime the spire-agent logs don't contain any suspicious activity that would show the agents would be busy at all.

Do you have debug logging enabled on the agent?

**Comment by szvincze** [comment]
No, debug logging is not enabled on this cluster. I have to ask my colleague to enable it and we will check. However on another cluster it is enabled, but the CPU is not that highly utilized there than on the one I attached the profile.

> this would imply that workloads are attesting, a lot.
How many attestations per seconds/minutes do you mean a lot?

**Comment by azdagron** [comment]
I don't know without seeing the logs, but for JSON parsing to show up that strongly on the profile, it would likely need to be in the order of hundreds a second I'd guess.

**Comment by szvincze** [comment]
> I don't know without seeing the logs, but for JSON parsing to show up that strongly on the profile, it would likely need to be in the order of hundreds a second I'd guess.

@azdagron: Do you need the whole logs or the information I provided is enough?

**Comment by evan2645** [comment]
Hi @szvincze , thanks for your patience.

First, based on your logs filling, it appears that your workloads are continually calling the workload API and hanging up. Any time the workload API receives a call, workload attestation is performed, and there is some cost to that. Not sure what libraries you're using, but workloads should generally not be doing this .. instead, they should hang onto the open stream, over which updates will be sent.

Second, we can see in the pprof that significant time is being spent in JSON parsing (and perhaps relatedly, GC). We do parse JSON response from the kubelet as part of k8s workload attestation, though we don't think the time spent there should be so significant. It's hard to tell if there's something about your environment tickling it, or if there is a bug or something in the JSON parsing logic. As a first step, please have a look at your workload behavior and see if we can correct things that way. Thanks again for your patience, please let us know what you find.

**Comment by szvincze** [comment]
Hi @evan2645,

Thanks for your comment.

Actually we use [grpc-health-probe](https://github.com/grpc-ecosystem/grpc-health-probe) for liveness and readiness probes of the workloads. It is spawning a new process and when option `-spiffe` is used, it triggers the k8s workload attestation periodically according to the frequency of the probes. Do you think that it is not a good practice to do this way?

Anyway, we will try to disable these probes to check whether it makes significant change in spire-agent's CPU utilization and come back with the outcome. However it is still strange that it happens in one sort of environments but we cannot observe in others.

**Comment by evan2645** [comment]
> Actually we use [grpc-health-probe](https://github.com/grpc-ecosystem/grpc-health-probe) for liveness and readiness probes of the workloads. It is spawning a new process and when option -spiffe is used, it triggers the k8s workload attestation periodically according to the frequency of the probes. Do you think that it is not a good practice to do this way?

This should be ok .. 4/sec is not a lot, and shouldn't cause this kind of load

Do you have cgroup memory limit applied to the agents? Are you doing any go gc tuning?

> However it is still strange that it happens in one sort of environments but we cannot observe in others.

It's surprising to spend so much time in the JSON decoder. One thing you could do here is hit the kubelet API in the affected cluster, get a copy of the JSON response that the agent would otherwise be receiving .. compare it to JSON response in a healthy cluster, and see if there are any obvious differences

**Comment by szvincze** [comment]
Hi @evan2645,  thanks for the hint.

Finally I got the JSON response from the `/pods` endpoint on the kubelet API from the nodes.
I created a table that contains the size of the response and the number of the pods in it, started with the one where we observed the highest CPU load and so on.
|Node ID| Size (Kb) | Num. pods|
|-----------|--------------|----------------|
|4           |2,317        |122             |
|5           |2,190        |103             |
|6           |1,679        |93               |
|7           |1,611        |77               |
|8           |1,133        |69               |

Is it possible that this size of the JSON the agent is parsing couple of times per second causes the high load?

**Comment by evan2645** [comment]
Hi @szvincze - yes , this could definitely be the problem.

We talked about this issue last week, and there are ~2 paths forward: try to further validate that this is in fact the problem (e.g. by putting together some Go benchmark tests and running your sample JSON through it to measure perf), or to try and patch it and see if it goes away. I think they may be similar amounts of effort, and this seems likely to be the problem, so we should probably just patch and see if we can improve.

We need only a subset of the data that is in the podspec. The standard library JSON parser is also apparently known to be not very performant. A patch here would involve first choosing a new JSON library that is performance oriented (e.g. [fastjson](https://github.com/valyala/fastjson)), and second, switching the [k8s workload attestor](https://github.com/spiffe/spire/blob/main/pkg/agent/plugin/workloadattestor/k8s/k8s.go#L589-L620) over to use it. We'd then want to deploy that patch into your environment and see if it resolves the problem.

If you're able to do this, we are happy to help guide you and get the fix merged if it works. If not, we will create a new issue to track the work, but it may take some time before we can find someone to pick it up.

**Comment by szvincze** [comment]
Hi @evan2645,  I ran the `fastjson` benchmarks with the data I collected from our clusters.
It shows significant difference compared to the standard JSON library. It should not be a problem to test the patched `spire-agent` on the problematic clusters but even if I would like to replace the library, I am not sure if I will have time recently to do it.

```
go test github.com/valyala/fastjson -bench='Parse$'
goos: linux
goarch: amd64
pkg: github.com/valyala/fastjson
cpu: 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
BenchmarkParse/u1/stdjson-map         	    26	  50066611 ns/op    47.38 MB/s	16994307 B/op	  277716 allocs/op
BenchmarkParse/u1/stdjson-struct      	    56	  20190732 ns/op   117.48 MB/s	     396 B/op	       6 allocs/op
BenchmarkParse/u1/stdjson-empty-struct      56	  20138404 ns/op   117.79 MB/s	     394 B/op	       6 allocs/op
BenchmarkParse/u1/fastjson                 304	   3812344 ns/op   622.20 MB/s	       0 B/op	       0 allocs/op
BenchmarkParse/u1/fastjson-get             313	   3770895 ns/op   629.04 MB/s	       0 B/op	       0 allocs/op
BenchmarkParse/s2/stdjson-map               25	  47296846 ns/op    45.25 MB/s	16034607 B/op	  253063 allocs/op
BenchmarkParse/s2/stdjson-struct            63	  18196694 ns/op   117.62 MB/s	     396 B/op	       6 allocs/op
BenchmarkParse/s2/stdjson-empty-struct      62	  18054551 ns/op   118.54 MB/s	     394 B/op	       6 allocs/op
BenchmarkParse/s2/fastjson                 336	   3581102 ns/op   597.65 MB/s	       0 B/op	       0 allocs/op
BenchmarkParse/s2/fastjson-get             332	   3574982 ns/op   598.67 MB/s	       0 B/op	       0 allocs/op
PASS
ok  	github.com/valyala/fastjson	13.159s
```

**Comment by szvincze** [comment]
@evan2645, it seems we definitely have issue with the high CPU load during workload attestation. Sometime it takes 10-20 minutes to get the worker node reloaded after restart because the agent cannot provide SVIDs to the pods.

Yesterday I had some time to look into it a bit deeper. So, I think I can create a patch and test it on the clusters where we observe the high CPU utilization.

As far as I see [fastjson](https://github.com/valyala/fastjson) does not use structs. It means a bit more changes in the k8s workload attestor not just switching it over to use it in the same way. Is it acceptable or we should keep the current structures?
